docs/service-implementation-101.md

```diff
@@ -0,0 +1,390 @@
+            &self.host,
+            self.port,
+            std::time::Duration::from_secs(5),
+            None,
+        ))?;
+
+        for t in self.subscriptions.lock().unwrap().iter() {
+            block_on(
+                self.mqtt
+                    .lock()
+                    .as_mut()
+                    .unwrap()
+                    .subscribe(t, QoS::AtMostOnce),
+            )?;
+        }
+
+        let msg = format!(
+            "{:?}",
+            String::from_utf8(
+                block_on(
+                    self.subscriber
+                        .lock()
+                        .as_mut()
+                        .unwrap()
+                        .as_mut()
+                        .unwrap()
+                        .recv()
+                )?
+                .payload
+            )
+        );
+
+        Ok(msg)
+    }
+}
+```
+
+We have mentioned dynamic dispatching — Let's do that now (in lib.rs`):
+```rs
+    fn pub_send_message_to_topic(
+        &mut self,
+        self_: &Self::Pub,
+        msg_key: PayloadParam<'_>,
+        msg_value: PayloadParam<'_>,
+        topic: &str,
+    ) -> Result<(), Error> {
+        match &self_.pub_implementor {
+            PubImplementor::ConfluentApacheKafka(pi) => {
+                pi.send_message_to_topic(msg_key, msg_value, topic)?
+            }
+            PubImplementor::Local(pi) => pi.send_message_to_topic(msg_key, msg_value, topic)?,
+        };
+
+// -snip-
+
+    fn sub_subscribe_to_topic(&mut self, self_: &Self::Sub, topic: Vec<&str>) -> Result<(), Error> {
+        match &self_.sub_implementor {
+            SubImplementor::ConfluentApacheKafka(si) => si.subscribe_to_topic(topic)?,
+            SubImplementor::Local(si) => {
+                si.subscribe_to_topic(topic.iter().map(|s| s.to_string()).collect::<Vec<String>>())?
+            }
+        }
+
+// -snip-
+    fn sub_poll_for_message(
+        &mut self,
+        self_: &Self::Sub,
+        timeout_in_secs: u64,
+    ) -> Result<Message, Error> {
+        Ok(match &self_.sub_implementor {
+            SubImplementor::ConfluentApacheKafka(si) => {
+                si.poll_for_message(timeout_in_secs)
+                    .map(|f| pubsub::Message {
+                        key: f.0,
+                        value: f.1,
+                    })?
+            }
+            SubImplementor::Local(si) => {
+                si.poll_for_message(timeout_in_secs)
+                    .map(|f| pubsub::Message {
+                        key: Some("batch".as_bytes().to_vec()),
+                        value: Some(f.as_bytes().to_vec()),
+                    })?
+            }
+        })
+    }
+```
+
+In here, all we are doing is adding calls to the functions implemented by the `MosquittoImplementor` under the `Local` variant of the `Pub/SubImplementor` enums, so that we are handling the case where a user provides a `toml` with `pubsub.mosquitto` and makes use of its' functionality.
```

Comment:
```suggestion
In here, all we are doing is adding calls to the functions implemented by the `MosquittoImplementor` under the `Local` variant of the `Pub/SubImplementor` enums, so that we are handling the case where a user provides a `toml` with `pubsub.mosquitto` and makes use of its functionality.
```